(RFC) WebUI mod

michaelortmann opened the pull request:
Found by: [michaelortmann](https://github.com/michaelortmann/)
Patch by: [michaelortmann](https://github.com/michaelortmann/)
Fixes: 

One-line summary:
WebUI mod adds webinterface to eggdrop

Additional description (if needed):
Screenshot desktop firefox: https://imgur.com/a/WONc3dR
Screenshot iOS safari: https://imgur.com/a/AQwxNgs (the imgur content detection/warning, is ridiculous ;)
Reuses eggdrops telnet interface
Adds minimal http/websocket
The html and [vanilla js](http://vanilla-js.com/) (best js framework ;) code is handcrafted
It sends our eggman as favicon :)
Work in progress, but already working for me
Eggdrops `make install` doesnt automatically copy from module to text directory, so you have to `cp src/mod/webui.mod/webui.html ~/eggdrop/text/` manually for now
Add to config:
```
loadmodule webui
listen 8080 webui
listen +8443 webui
```
Yes, the `+` means it already supports https

Test cases demonstrating functionality (if applicable):
Tested with destop firefox, desktop brave (based on chromium) and iphone iOS safari

michaelortmann committed:
Initial commit

Commit:
Merge remote-tracking branch 'upstream/develop' into webui.mod

Commit:
Merge remote-tracking branch 'upstream/develop' into webui.mod

Commit:
Merge remote-tracking branch 'upstream/develop' into webui.mod

Commit:
Merge remote-tracking branch 'upstream/develop' into webui.mod

Commit:
Enhance log

Commit:
Remove debug log

Commit:
Update Copyright

Commit:
Renamed webui.html -> text/webui.html

Commit:
add character encoding

Commit:
Remove trailing slash on void element

Commit:
Use CSS instead of obsolete elements

Commit:
Add lang attribute

vanosg commented:
```
18:05:47] Telnet connection: localhost/46032
[18:05:47] TLS: handshake successful. Secure connection established.
[18:05:47] TLS: peer did not present a certificate
[18:05:47] Denied telnet: *, No Access

Thread 1 "eggdrop" received signal SIGSEGV, Segmentation fault.
0x00007ffff7d81aca in ?? () from /lib/x86_64-linux-gnu/libssl.so.3
(gdb) bt
#0  0x00007ffff7d81aca in ?? () from /lib/x86_64-linux-gnu/libssl.so.3
#1  0x00007ffff7d81cb7 in SSL_read () from /lib/x86_64-linux-gnu/libssl.so.3
#2  0x00007ffff76b33b5 in webui_http_activity (idx=7, 
    buf=0x7fffffffbde1 "P/7Es4T4NIGlFL7i8kbrrw==\r\nConnection: keep-alive, Upgrade\r\nCookie: csrftoken=JEqDnapUmFU1zeZwSab3r0uzcb885hlU; sessionid=kexs0xja1ymwhumezcss3p4doaabi42s; messages=W1siX19qc2", len=511) at .././webui.mod/webui.c:333
```

and a second one with debug enabled
```
<snip web code>
[18:11:59] flags 1026
[18:11:59] webui: webui_http_activity(): idx 6 len 511
[18:11:59] webui: http()
[18:11:59] webui: GET /w
[18:11:59] webui: server requests websocket upgrade
[18:11:59] webui: tputs(): >>>HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: ljLkHJvOkjthVo/X5aPf0zQd3MM=

<<< 129
[18:11:59] webui: unset flag SOCK_BINARY sock 18

[18:11:59] webui: set flag SOCK_WS socklist 6 idx 6 sock 18 status 0

[18:11:59] webui: found (webui) dcc

[18:11:59] Denied telnet: *, No Access
[18:11:59] Received close notify warning during write
[18:11:59] webui: CHANGEOVER -> idx 6 sock -1


Thread 1 "eggdrop" received signal SIGSEGV, Segmentation fault.
0x00007ffff7d81aca in ?? () from /lib/x86_64-linux-gnu/libssl.so.3
(gdb) bt
#0  0x00007ffff7d81aca in ?? () from /lib/x86_64-linux-gnu/libssl.so.3
#1  0x00007ffff7d81cb7 in SSL_read () from /lib/x86_64-linux-gnu/libssl.so.3
#2  0x00007ffff76b33b5 in webui_http_activity (idx=6, 
    buf=0x7fffffffbde1 "iqubpCvpepPdkRKCJVsTIw==\r\nConnection: keep-alive, Upgrade\r\nCookie: csrftoken=JEqDnapUmFU1zeZwSab3r0uzcb885hlU; sessionid=kexs0xja1ymwhumezcss3p4doaabi42s; messages=W1siX19qc2", len=511) at .././webui.mod/webui.c:333
#3  0x00005555555a9f49 in mainloop (toplevel=1) at ./main.c:853
```

and a third non-crash, but still can't connect over a TLS port with https:
```
</html>
<<< 1539
[18:40:41] webui: SSL_read(): len 188
[18:40:41] webui: webui_http_activity(): user 0.108ms sys 0.100ms
[18:40:41] net: connect! sock 7
[18:40:41] sockread sock 7 return 4 >><< 0 binary? 0
[18:40:41] flags 4
[New Thread 0x7ffff6dff640 (LWP 2121829)]
[18:40:41] dcc_telnet_hostresolved()
[Thread 0x7ffff6dff640 (LWP 2121829) exited]
[18:40:41] Telnet connection: localhost/48796
[18:40:41] TLS: attempting SSL negotiation...
[18:40:41] TLS: state change: before SSL initialization
[18:40:41] TLS: state change: before SSL initialization
[18:40:41] TLS: awaiting more reads
[18:40:41] TLS: handshake in progress
[18:40:41] webui_dcc_telnet_hostresolved(6)
[18:40:41] TLS: state change: before SSL initialization
[18:40:41] TLS: alert during write: fatal (internal error).
[18:40:41] TLS: awaiting more writes
[18:40:41] net: sockread(): SSL_read() error = error:0A000115:SSL routines::session id context uninitialized (1)
[18:40:41] net: eof!(read) socket 11
[18:40:41] webui: webui_http_eof() idx 6 sock 11
```

Commit:
Fix for sock == -1 if lostdcc() in dcc_telnet_hostresolved2()

Commit:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann commented:
@vanosg: The crash should be fixed now.

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
Fix proto.h / --disable-tls

michaelortmann committed:
Enhance TODO doc

michaelortmann committed:
Fix --disable-tls

vanosg commented:
One more- 
```/home/runner/work/eggdrop/eggdrop/eggdrop/src/tls.c:233: undefined reference to `SSL_CTX_get0_certificate'```
in OpenSSL 0.9.8

michaelortmann committed:
Fix openssl 0.9.8 even for WIP/testing

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann commented:
> One more- `` /home/runner/work/eggdrop/eggdrop/eggdrop/src/tls.c:233: undefined reference to `SSL_CTX_get0_certificate' `` in OpenSSL 0.9.8

Yes, this is an unfinised block of code, waiting for #1411. Once #1411 is merged ill finish it. Until then, for now, i added an `ifdef` around it.

michaelortmann committed:
Fix openssl version < 1.1

michaelortmann committed:
misc/makedepend

michaelortmann committed:
Added experimental http feedback, if user connects http to https port

michaelortmann committed:
misc/makedepend

michaelortmann committed:
Fix format specifier

michaelortmann committed:
Remove forgotten debug

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
Cleanup

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
const

michaelortmann committed:
Cleanup

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
Cleanup

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
Add html header Content-Type

michaelortmann committed:
Add stealth telnets

michaelortmann committed:
Bump version number

michaelortmann committed:
Cleanup

michaelortmann committed:
Fix merge

michaelortmann committed:
Add naive escape_html()

michaelortmann committed:
Cleanup webui_frame()

michaelortmann committed:
Enhance webui_frame() API

michaelortmann committed:
Cleanup

michaelortmann committed:
Use Prettier as a code formatter

michaelortmann committed:
Cleanup, remove debug log

michaelortmann committed:
Handle connection closed by foreign host

michaelortmann committed:
Update

michaelortmann committed:
Unset webui http socket option SOCK_BUFFER

michaelortmann committed:
Fix --disable-tls

michaelortmann committed:
Cleanup

michaelortmann committed:
Cleanup

michaelortmann committed:
Enhance eggdrop.conf

michaelortmann committed:
Enhance code comment

michaelortmann committed:
Cleanup

michaelortmann committed:
Update module dependency

michaelortmann committed:
Cleanup

michaelortmann committed:
Dyn alloc response

michaelortmann committed:
Cleanup

michaelortmann committed:
Add favicon as file and enhance compression using optipng

michaelortmann committed:
Fix webui.html again

michaelortmann committed:
Fix traffic counting

michaelortmann committed:
Fix uninitialised value, see #1845

michaelortmann committed:
Cleanup

michaelortmann committed:
Cherry pick #1846

michaelortmann committed:
New favicon.ico and apple-touch-icon.png

Put files to webui/

Raise telnet-flood to 16:60 because web clients connect more than once

michaelortmann committed:
Re-add console log foer echo on/off

michaelortmann committed:
Add 404 and fix indent

michaelortmann committed:
Enhance 404

michaelortmann committed:
Cleanup

michaelortmann committed:
Cache file data by modification time

michaelortmann committed:
Cleanup

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
Fix module unloading

michaelortmann committed:
Log sock instead if idx

michaelortmann committed:
Cleanup

michaelortmann committed:
Add comment

michaelortmann committed:
Cleanup and comment

michaelortmann committed:
Indent, also add temp. assert while this PR is in development

michaelortmann committed:
Bufsize

michaelortmann committed:
Fix telnet echo and esc bold codes

michaelortmann committed:
Fix bold - thank you CrazyCat

michaelortmann committed:
Fix format specifier for off_t

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
Update Makefile.in

michaelortmann committed:
Fix tls

michaelortmann committed:
Cleanup

michaelortmann committed:
Merge remote-tracking branch 'upstream/develop' into webui.mod

michaelortmann committed:
Cleanup

michaelortmann committed:
Disable webui mod by default
